[addons] change and cleanup binary addon parts

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This is another step to rework the binary addon system and prepare for further changes where needed parts for bigger binary addon's (e.g. PVR) becomes improved.

On them here is mainly no more needed parts removed and a function added to get single CAddonInfo, used in future in case of reported changes of CAddonManager to get updated infos.

With the obsolete marked ADDON_STATUS_NEED_SAVEDSETTINGS. This becomes removed in future, if removed now need every addon updated.

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons] remove the use of ADDON_STATUS_NEED_SAVEDSETTINGS

The way how the addon has saved the changed settings with them was a poor
hack way and from all addon developers only seen with a "?".

This remove them. It can be also done by a "kodi::SetSetting..." callback.

AlwinEsch committed:
[addons] remove the dll copy hack way

This remove the for audio decoder addons used way where the related
dll was copied to another name to allow open again for another work.

This becomes in done now from the addons itself.

AlwinEsch committed:
[addons] remove dead defined functions in CAddonDll header

FernetMenta reviewed (commented):
I trust you :)

AlwinEsch commented:
Thanks @FernetMenta :-D

AlwinEsch committed:
[addons] fix compile warning

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch committed:
[addons] remove not needed includes on AddonManager

Hope to fix the Mac fault with them.

ksooo commented on xbmc/addons/AddonManager.h:
```diff
@@ -138,6 +134,18 @@ namespace ADDON
+     * @brief To get requested installed binary addon on Kodi
+     *
+     * This function becomes used from ADDON::CBinaryAddonManager to get his
+     * related addon (whether enabled or disabled).
```
This function is used by ADDON::CBinaryAddonManager to obtain the add-on with the given id, regardless the add-on is disabled or enabled.

AlwinEsch replied:
Becomes changed :)

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -82,7 +82,9 @@ typedef enum ADDON_STATUS
+  /// @warning obsolete and no more used from Kodi!
   ///
+  /// Use kodi::SetSetting... during end to store changed settings
   ADDON_STATUS_NEED_SAVEDSETTINGS,
```
If it is no longer used, remove it, please. Yes I know that some addons need to be bumped for this, but still.

AlwinEsch replied:
Then I add a also a another commit where other functions becomes added and need a complete version change :D

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Next try :-(, crrrrrrrrrrr

Can someone explain me this error:
```
In file included from /Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/AddonManager.cpp:21:
In file included from /Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/AddonManager.h:22:
In file included from /Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/Addon.h:22:
In file included from /Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/IAddon.h:29:
In file included from /Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/AddonInfo.h:22:
In file included from /Users/Shared/jenkins/workspace/OSX-64/xbmc/XBDateTime.h:25:
In file included from /Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/string:439:
In file included from /Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:626:
/Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:294:15: error: no viable overloaded '='
        first = __p.first;
        ~~~~~ ^ ~~~~~~~~~
/Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:374:20: note: in instantiation of member function 'std::__1::pair<const ADDON::AddonVersion, bool>::operator=' requested here
            second = _VSTD::forward<_U1>(_VSTD::get<1>(__p));
                   ^
/Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:610:15: note: in instantiation of function template specialization 'std::__1::pair<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >::operator=<const std::__1::pair<const std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> > &, void>' requested here
        {__nc = __v.__cc; return *this;}
              ^
/Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:1263:35: note: in instantiation of member function 'std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >::operator=' requested here
                __cache->__value_ = *__first;
                                  ^
/Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:1204:9: note: in instantiation of function template specialization 'std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >, std::__1::__map_value_compare<std::__1::basic_string<char>, std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >, std::__1::less<std::__1::basic_string<char> >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> > > >::__assign_multi<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >, void *> *, long> >' requested here
        __assign_multi(__t.begin(), __t.end());
        ^
/Applications/Xcode6.1.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:885:21: note: in instantiation of member function 'std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >, std::__1::__map_value_compare<std::__1::basic_string<char>, std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> >, std::__1::less<std::__1::basic_string<char> >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> > > >::operator=' requested here
            __tree_ = __m.__tree_;
                    ^
/Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/AddonInfo.h:91:9: note: in instantiation of member function 'std::__1::map<std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool>, std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, std::__1::pair<const ADDON::AddonVersion, bool> > > >::operator=' requested here
  class CAddonInfo
        ^
/Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/AddonVersion.h:70:38: note: candidate function not viable: 'this' argument has type 'const ADDON::AddonVersion', but method is not marked const
  inline AddonVersion& AddonVersion::operator=(const AddonVersion& other)
                                     ^
1 error generated.
make[2]: *** [build/addons/CMakeFiles/addons.dir/AddonManager.cpp.o] Error 1
make[1]: *** [build/addons/CMakeFiles/addons.dir/all] Error 2
make: *** [all] Error 2
```
Others have tested it on their Mac and it went.
Also it works on all other OS.

Comes from [	[addons] add possibility to get single CAddonInfo if needed](https://github.com/xbmc/xbmc/pull/12410/commits/73d7cbe8c9fc5f0afd81d110da8dd05a0d2276de)

Has added several other commits behind to find a way to fix jenkins. But still not successful :(

Hope with the last it works but not want to touch `ADDON::AddonVersion` who bring a problem.

Does not make sense for me if one causes the other. Although one has nothing to do with the other.

AlwinEsch commented:
jenkins build this please

ksooo commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -566,26 +566,26 @@ bool CAddonMgr::GetInstalledBinaryAddon(const std::string& addonId, BINARY_ADDON
+//       }
+//     }
+//     m_cpluff->release_info(m_cp_context, cp_addon);
+//   }
```
Please remove this. Not actually loosing stuff is what we have git for.

ksooo commented on xbmc/addons/Addon.h:
```diff
@@ -19,15 +19,15 @@
+// // #include "utils/ISerializable.h"
+//
+// #include <memory>
+// #include <vector>
```
please remove, instead of commenting out

AlwinEsch replied:
Is only on test :-D, becomes removed

ksooo commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -577,27 +551,43 @@ bool CAddonMgr::GetInstalledBinaryAddons(BINARY_ADDON_LIST& binaryAddonList)
+//       if (!value.empty() &&
+//           props->plugin->plugin_path &&
+//           strcmp(props->plugin->plugin_path, "") != 0 &&
+//           Factory(cp_addon, ADDON_UNKNOWN, builder, true))
```
hmm. why are you commenting out instead of actually removing code? If it is a test, please don't push.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch committed:
[addons] remove not needed included

AlwinEsch committed:
[addons] add possibility to get single CAddonInfo if needed

This is needed on binary addon manager to get addon info in case
of changes. Without must be the whole addon list loaded what make
no sense in case by only one changed addon.

AlwinEsch committed:
[addons] remove no more used ADDON_STATUS_NEED_SAVEDSETTINGS and version increase

WARNING: All addons need a recompile!

AlwinEsch commented:
@ksooo is updated and the ADDON_STATUS_NEED_SAVEDSETTINGS complete removed.

garbear commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -66,6 +75,53 @@ void Interface_General::DeInit(AddonGlobalInterface* addonInterface)
+  }
+
+  std::string str;
+  if (strcmpi(id, "author") == 0)
```
missing `#include <string.h>`

AlwinEsch replied:
is added

AlwinEsch committed:
[addons] add other useful general addon callback functions

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Is in.

As note to all @xbmc/team-kodi, binary addons need recompile to work!
